torchdynamo/eval_frame.py

```diff
@@ -397,7 +398,88 @@ def guard_export_print(guards):
     return explanation, out_guards, graphs, ops_per_graph
 
 
-def export(f, *args, aten_graph=False, **kwargs):
+class PartiallyDynamicTensor(torch.Tensor):
```

Comment:
I don't like the idea of using Tensor subclassing for this.  I think it makes the UX a bit awkward.

One option is we already have a similar concept in `torch.fx.tensor_type.*`.  See `test_gradual_type.py` in the pytorch repo.

That uses type annotations, like:
```
from torch.fx.tensor_type import TensorType, Dyn

def forward(self, x: TensorType((1, 2, 3, Dyn)), y: Dyn):
   ...
```

Reply:
Hmm, I kind of prefer keeping them as Tensors. Why do you not like Tensor subclassing for this? What makes the UX awkward?

Reply:
This is yet another way to declare what dims are static versus dynamic when we already have a few different other ones.

The other one that comes to mind is `torch.onnx.export(..., dynamic_axes={"x": [2,3]})`

`torch.fx.tensor_type.TensorType` brings with is shape propagation.

Reply:
The onnx way looks inferior to this (Matching names of tensors to dynamic axes in a dict feels brittle).

For the other approach, where we use`torch.fx.tensor_type.TensorType`... Changing to`TensorType(num1, num2, num3, dyn)` *in the signature* 

(A) feels less adaptable and like worse UX than taking your already existing inputs and just passing them in as composed through `PartiallyDynamicTensor.create(yourTensor, dynamic_dims=[3])`. 

(B) (If I understood it correctly) Requires changing the model, where callers of .export may often not be able to do that, or if they are, may not have a consistently uniform pattern to the dynamism they want to add (or constraints they want to introduce).

I think this actually makes for a less awkward UX. 

That being said, if you really dislike the idea of this being a tensor subclass, we can make it compositional instead.

Reply:
(A) The UX seems roughly the same, though this is adding yet another way to do something (annotate static/dynamic shapes) we already have solved in many other places. It is fragmentation for no real benefit.

I personally like type annotations better.  This subclass doesn't change the Tensor's behavior in eager mode, so it feels more like an annotation than a subclass.

(B) If you can't change the model, just add a wrapper function with annotations.

The huge problem both of these are going to have is propagation.  The annotations and the subclasses both only apply to the program inputs, and we don't have a good way or rules of propagating them through the program.

Reply:
> we already have solved in many other places.

I see that we solved it really just once, with `torch.fx.tensor_type.TensorType`. Once is enough, but perhaps not many? 

> If you can't change the model, just add a wrapper function with annotations.

what if populating the value needs to be dynamic? Does one need to come up with new annotation functions? What if which dim is dynamic is not determined at authoring time? 

Maybe this is contrived, but philosophically it feels like the constraint is on the tensor as an input, not on the definition of the model/function.

> The annotations and the subclasses both only apply to the program inputs, and we don't have a good way or rules of propagating them through the program.

That is fair, but is there value in such a propagation for our use case?

Reply:
> Maybe this is contrived, but philosophically it feels like the constraint is on the tensor as an input, not on the definition of the model/function.

That is our use case in Torch-MLIR, fwiw. Although it does work well to define a wrapper function/class with the right annotations and then script/trace/td.export it. Introducing a new "tensor with partially dynamic shape data attached" does seem like an extra concept here.

> That is fair, but is there value in such a propagation for our use case?

This propagation is exactly what Torch-MLIR would like to get here. Currently we maintain our own shape propagation infrastructure, thankfully sharing @eellison's python definitions. The goal of this feature request is for torchdynamo.export to just give us the maximally precise shapes propagated through the entire program without us having to maintain our own shape propagation infrastructure. Is there a plan for this?

Although technically Torch-MLIR takes JIT IR as input, so I think that in principle even if torchdynamo gave the partially propagated shapes, the going down to JIT IR could lose them. Any thoughts on how best to get from TD FX graphs down to JIT IR while retaining partially specified shapes?

cc @Chillee

Reply:
> Is there a plan for this?

Plan? No.

Desire? Yes.

I spoke to @Chillee about this today - he mentioned FakeTensor is the closest thing we have here w/r/t the propagation behavior we want. 

I plan to start poking around this week at symbolic shapes and some of the work that @Chillee has already done there to see how far it can take us, and where the gaps are

Reply:
@ezyang has a [symbolic shapes branch](https://github.com/pytorch/pytorch/pull/84246) that would allow symbolic shapes to propagate through the dispatcher using FakeTensor.

It is not ready for use yet, but at some point I'd hope it would give us shape propagation.